app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php

```diff
@@ -90,18 +99,37 @@ public function aroundSave(ItemResourceModel $subject, callable $proceed, Abstra
     }
 
     /**
-     * @param $legacyStockItem
+     * @param AbstractModel $legacyStockItem
      *
      * @return string
      */
-    private function getTypeId($legacyStockItem): string
+    private function getTypeId(AbstractModel $legacyStockItem): string
     {
         $typeId = $legacyStockItem->getTypeId();
-        if (null === $typeId) {
-            $product = $this->productRepository->getById($legacyStockItem->getProductId());
-            $typeId = $product->getTypeId();
+        if ($typeId === null) {
+            $sku = $legacyStockItem->getSku();
+            if ($sku === null) {
+                $sku = $this->getProductSkuById((int)$legacyStockItem->getProductId());
+            }
+            $typeId = $this->getProductTypeBySku->execute([$sku])[$sku];
         }
 
         return $typeId;
     }
+
+    /**
+     * Get product sku by id
+     *
+     * @param int $productId
+     * @return string|null
```

Comment:
We don't have cases when the method returns `null`
See implementation of interface `\Magento\InventoryCatalog\Model\GetProductIdsBySkusInterface`

Reply:
fixed